### PR TITLE
[DS-4522] Logical item filtering (with DOI implementation)

### LINKED ISSUE #7855: [DS-4522] Logical item filtering (with implementation of filtered DOI provider)
<i>Imported from JIRA <a href="https://jira.lyrasis.org/browse/DS-4522">[DS-4522]</a> created by kshepherd</i>
<p>This feature allows selective filtering of items for DOI minting, as well as ways for administrators to override/skip those filters, and brings with it a complete generic logical filtering system for DSpace items (similar in concept to the XOAI XML filter definitions) that can be used by any other service or component of DSpace.</p>

<p>It was developed by The Library Code GmbH with the support of Technische Informationsbibliothek TIB and the Technische Universität Hamburg TUHH.</p>

<p>Full Documentation (wiki page): <a href="https://wiki.lyrasis.org/display/~kshepherd/DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.lyrasis.org/display/~kshepherd/DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7</a></p>

<p>PR: <a href="https://github.com/DSpace/DSpace/pull/2789" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/DSpace/DSpace/pull/2789</a></p>

### PR DESCRIPTION
**Pull request by kshepherd**
## References
* JIRA: [DS-4522 JIRA ticket](https://jira.lyrasis.org/browse/DS-4522) ticket
* Fixes #7855 
* Documentation: [Wiki draft documentation page](https://wiki.lyrasis.org/pages/viewpage.action?spaceKey=~kshepherd&title=DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7)

## Description
This feature allows selective filtering of items for DOI minting, as well as ways for administrators to override/skip those filters, and brings with it a complete generic logical filtering system for DSpace items (similar in concept to the XOAI XML filter definitions) that can be used by any other service or component of DSpace.
A more thorough explanation is at [the draft wiki page](https://wiki.lyrasis.org/pages/viewpage.action?spaceKey=~kshepherd&title=DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7)

## Instructions for Reviewers
List of changes in this PR:
* Addition of classes in org.dspace.content.logic package, representing Logical Statements used in filtering
* Refactor of DOI Providers to make use of optional filters
* New _item-filters.xml_ spring configuration for defining item filters, with some examples already included
* Unit tests for logical filters, and updated tests for DOI providers to test filtering of items in DOI events
* Test command-line client for trying out filters without requiring them in production configuration

### To test: ###
* Have a read a of the [doco](https://wiki.lyrasis.org/pages/viewpage.action?spaceKey=~kshepherd&title=DS-4522+Logical+Item+Filtering+and+DOI+Filtered+Provider+for+DSpace+7) and get a feeling for how the filter spring configuration works, the difference between filters, operators, conditions, etc.
* Take a look at `item-filters.xml` and look for some filters to test, or define your own
* Use the command-line client to test out the filter over some items you know will pass or fail:
`dsrun org.dspace.content.logic.TestLogicRunner -f FILTERNAME -i HANDLE`
* Enable DOI providers / connectors the usual way, and include the filter in `identifier-service.xml` (there is a commented-out example)
* Test things normal operation with the DOI consumer, CLI tool, and make sure the filters are applied (unless explictly skipped with the CLI tool)
* Test out the curation task `registerdoi`, which by default should **always** bypass any filtering. It can be configured to obey filters instead, and to distribute across collections. (see `config/modules/doi-curation.cfg`)

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is not small, at well over 1000 lines long including the new classes, and DOI provider refactors. I hope an exception can be made for this new feature.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR includes no 3rd party dependencies
- [x] My PR does not modify REST contract or Angular UI

## Credits
This PR was developed by The Library Code GmbH with the support of Technische Informationsbibliothek (TIB) and the Technische Universität Hamburg TUHH.

### CODE REVIEW

**Comment by kshepherd** [comment]
I apologise, my last rebase included some inadvertant commits, and I also have some licence headers to add. I will close this PR and open a new one tomorrow.

**Comment by kshepherd** [comment]
(I resolved the recent conflict raised when UI tasknames were removed from curate.cfg)

File: dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    private DOIIdentifierProvider provider;
+
+    /**
+     * Initialise the curation task and read configuration, instantiate the DOI provider
+     */
+    @Override
+    public void init(Curator curator, String taskId) throws IOException {
+        super.init(curator, taskId);
+        // Get 'skip filter' behaviour from configuration, with a default value of 'true'
+        skipFilter = configurationService.getBooleanProperty(PLUGIN_PREFIX + ".skip-filter", true);
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java:50]: I would prefer that to default to false

File: dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java [hunk 2]
```diff
@@ -671,8 +885,31 @@ public String getDOIByObject(Context context, DSpaceObject dso)
...
+     *
+     * @param context       - DSpace context
+     * @param dso           - DSpaceObject to identify
+     * @param doiIdentifier - DOI to load or create (null to mint a new one)
+     * @param skipFilter    - Whether or not to skip the filters for the canMint() check
+     * @return
+     * @throws SQLException
+     * @throws DOIIdentifierException
+     */
+    protected DOI loadOrCreateDOI(Context context, DSpaceObject dso, String doiIdentifier, Boolean skipFilter)
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java:907]: better to use a primitive boolean

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Register a new identifier for a given DSpaceObject
+     * @param context    - DSpace context
+     * @param dso        - DSpaceObject to use for identifier registration
+     * @param skipFilter - boolean indicating whether to skip any filtering of items before performing registration
+     * @return identifier
+     * @throws IdentifierException
+     */
+    public abstract String register(Context context, DSpaceObject dso, Boolean skipFilter)
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:46]: better to use a primitive boolean

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Register a specified identifier for a given DSpaceObject
+     * @param context    - DSpace context
+     * @param dso        - DSpaceObject identified by the new identifier
+     * @param identifier - String containing the identifier to register
+     * @param skipFilter - boolean indicating whether to skip any filtering of items before performing registration
+     * @throws IdentifierException
+     */
+    public abstract void register(Context context, DSpaceObject dso, String identifier, Boolean skipFilter)
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:57]: better to use a primitive boolean

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+     * Reserve a specified identifier for a given DSpaceObject (eg. reserving a DOI online with a registration agency)
+     * @param context    - DSpace context
+     * @param dso        - DSpaceObject identified by this identifier
+     * @param identifier - String containing the identifier to reserve
+     * @param skipFilter - boolean indicating whether to skip any filtering of items before performing reservation
+     * @throws IdentifierException
+     * @throws IllegalArgumentException
+     * @throws SQLException
+     */
+    public abstract void reserve(Context context, DSpaceObject dso, String identifier, Boolean skipFilter)
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:70]: better to use a primitive boolean

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Mint a new identifier in DSpace - this is usually the first step of registration
+     * @param context    - DSpace context
+     * @param dso        - DSpaceObject identified by the new identifier
+     * @param skipFilter - boolean indicating whether to skip any filtering of items before minting
+     * @return a String containing the new identifier
+     * @throws IdentifierException
+     */
+    public abstract String mint(Context context, DSpaceObject dso, Boolean skipFilter) throws IdentifierException;
```
Comment (abollini) [inline dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:81]: better to use a primitive boolean

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+     */
+    @After
+    @Override
+    public void destroy() {
+        context.turnOffAuthorisationSystem();
+        // Delete resources
+        try {
+            itemService.delete(context, itemOne);
+            itemService.delete(context, itemTwo);
+            itemService.delete(context, itemOne);
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:133]: this look to be a duplicate code that could lead to an ignored NPE in cleanup

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+            collectionService.delete(context, collection);
+            communityService.delete(context, owningCommunity);
+        } catch (Exception e) {
+            // ignore
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:137]: better to log as it could help to trouble shooting issue in test cleanup

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    /**
+     * Test the AND operator with simple lists of logical statements
+     */
+    @Test
+    public void testAndOperator() {
+        // Blank operator
+        And and = new And();
+        // Try tests - the item can be null, as the statements are simply returning booleans themselves
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:165]: I disagree here as null is a very special value and we are hardcoding in the test the logic of our current implementation. Any statement about a null item could/should return false.
Can you switch the test to use a real item?

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+     * Test a simple filter with a single logical statement: the MetadataValueMatchCondition
+     * looking for a dc.title field beginning with "TEST"
+     */
+    @Test
+    public void testMetadataValueMatchCondition() {
+        try {
+            MetadataValue metadataValueOne = metadataValueService.create(context, itemOne, metadataField);
+            MetadataValue metadataValueTwo = metadataValueService.create(context, itemTwo, metadataField);
+            metadataValueOne.setValue("TEST title should match the condition");
+            metadataValueTwo.setValue("This title should not match the condition");
```
Comment (abollini) [inline dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:299]: a test with an item without the checked metadata should be included

**Review by abollini — CHANGES_REQUESTED** [review]
hi @kshepherd 
thanks for this very useful contribution. I like the idea and the code looks great overall. Right now I have made only a code inspection and checked the documentation, I have added few inline comments and i would appreciate if you can add Test for a couple of Condition classes that seems to be not covered by any test
- BitstreamCountCondition
- InCollectionCondition
- InCommunityCondition
- MetadataValuesMatchCondition (only the single from is tested)
- ReadableByGroupCondition

once this feedback is addressed and the conflict solved I can run a complete test and finalize the review.

File: dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java (hunk 1 above)
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java:50]: I haven't looked deep into this and might be wrong. But I think "true" is the conservative way here. DSpace is not able to filter DOIs so far. The filter is new. Skipping the filter is the old behavior. Do I miss anything?

**Comment by kshepherd** [comment]
I've discovered a bug in the InCollectionCondition (needs to use dso.getParentObject() rather than collections or owningCollection()), I will rebase and push a fix soon and carry on with tests

File: dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+package org.dspace.content.logic;
+
+/**
+ * Exception for errors encountered while evaluating logical statements
+ * defined as spring beans.
+ *
+ * @author Kim Shepherd
+ * @version $Revision$
+ */
+public class LogicalStatementException extends Exception {
```
Comment (LucaGiamminonni) [inline dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java:17]: hi @kshepherd, what about making this exception unchecked rather than checked? In this way the callers of the methods that throw this exception would not be forced to handle it all the time, since it might not even be clear what it is due to and how to handle it. In a more general discourse, I would prefer to use unchecked exception instead of checked because they create problems with the use of lambdas and create a coupling between the different service layers of the system. What do you think about it?

File: dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java [hunk 12]
```diff
@@ -832,7 +1089,39 @@ protected void removeDOIFromObject(Context context, DSpaceObject dso, String doi
...
+     * @param context
+     * @param dso The item to be evaluated
+     * @return
+     * @throws DOIIdentifierNotApplicableException
+     */
+    @Override
+    public Boolean canMint(Context context, DSpaceObject dso) throws DOIIdentifierNotApplicableException {
+        // Default is 'true' in the case of a null/missing filter. All we really care about is whether
+        // an exception was thrown or not.
+        log.debug("canMint is being called");
```
Comment (corrad82-4s) [inline dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java:1107]: Hi @kshepherd , sorry to jump into this discussion.
I understand the expected behaviour of this method as correctly explained in comments, and agree on what we really care about. 
But with boolean as returned value, what happens if a client of this method, unaware of its behavior, calls it and makes his/her logic or whatever by handling a false return value that actually is not returned? 
Thank you

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java (hunk 6 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:81]: see my other comment near the canMint section - a null or missing filter should not default to false, so I didn't want to use a primitive here

File: dspace-api/src/main/java/org/dspace/content/logic/Filter.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+ */
+public interface Filter extends LogicalStatement {
+    /**
+     * Get the result of logical evaluation for an item
+     * @param context   DSpace context
+     * @param item      Item to evaluate
+     * @return boolean
+     * @throws LogicalStatementException
+     */
+    Boolean getResult(Context context, Item item) throws LogicalStatementException;
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/logic/Filter.java:34]: see my other comment near the canMint section - i didn't want to automatically default to false or anything other than true - it might be needed to handle nulls for some better error handlig that doesn't cause unexpected logical results

File: dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java (hunk 11 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java:17]: Thanks @LucaGiamminonni , I see what you mean. I'll think about how I can refactor this, maybe best made after some of the other suggested changes

File: dspace-api/src/main/java/org/dspace/content/logic/condition/AbstractCondition.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    public Boolean getResult(Context context, Item item) throws LogicalStatementException {
+        if (item == null) {
+            throw new LogicalStatementException("Item is null");
+        }
+        if (context == null) {
+            throw new LogicalStatementException("Context is null");
+        }
+        if (this.parameters == null) {
+            throw new LogicalStatementException("Parameters are null");
+        }
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/logic/condition/AbstractCondition.java:76]: thanks, i'll think about this

File: dspace-api/src/main/java/org/dspace/content/logic/condition/ReadableByGroupCondition.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * A condition that accepts a group and action parameter and returns true if the group
+ * can perform the action on a given item
+ *
+ * @author Kim Shepherd
+ * @version $Revision$
+ */
+public class ReadableByGroupCondition extends AbstractCondition {
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/logic/condition/ReadableByGroupCondition.java:29]: yes, good suggestion

File: dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java (hunk 1 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java:50]: yes @abollini has left a preference here and in the configuration of a default to false.
i do think 'true' is the conservative approach and matches existing DSpace operation.
another way to look at it is "skipFilter = true" is the same as "alwaysApplyFilter = false", if that makes sense?
this option is only used in the CLI, curation task, special minting buttons in the UI which allows administrators to force through DOIs that would otherwise be filtered out in normal operation

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java (hunk 3 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:46]: see my other comment near the canMint section - a null or missing filter should not default to false, so I didn't want to use a primitive here

@abollini perhaps this also means i have confused things iwth my naming... would it be clearer code to read if there wa a primitive boolean called `applyFilter` instead of `skipFilter`? then it would be save since a null or missing filter would mean the default behaviour was applyFilter == false which is good.
currently, if skipFilter is a primitive and ends up false for some non-logical reason (null or missing filter, some other error causes null result to be returend) then the filter is still applied and that would definitely cause problems.

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java (hunk 4 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:57]: see my other comment near the canMint section - a null or missing filter should not default to false, so I didn't want to use a primitive here

perhaps this also means i have confused things iwth my naming... would it be clearer code to read if there wa a primitive boolean called applyFilter instead of skipFilter? then it would be save since a null or missing filter would mean the default behaviour was applyFilter == false which is good.
currently, if skipFilter is a primitive and ends up false for some non-logical reason (null or missing filter, some other error causes null result to be returend) then the filter is still applied and that would definitely cause problems.

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java (hunk 5 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:70]: see my other comment near the canMint section - a null or missing filter should not default to false, so I didn't want to use a primitive here

perhaps this also means i have confused things iwth my naming... would it be clearer code to read if there wa a primitive boolean called applyFilter instead of skipFilter? then it would be save since a null or missing filter would mean the default behaviour was applyFilter == false which is good.
currently, if skipFilter is a primitive and ends up false for some non-logical reason (null or missing filter, some other error causes null result to be returend) then the filter is still applied and that would definitely cause problems.

File: dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java (hunk 12 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java:1107]: Hi @corrad82-4s , I see what you're getting at... what would be your suggestion? Should we make this methoid void instead of Boolean and perhaps make it clearer by renaming to `checkMintable` ?

Comment (corrad82-4s) [reply dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java:1107]: hi @kshepherd , 
yes, if there is no need to deal with returned value, and the purpose of this method is raise an exception if a dso is not mintable I would make it void and rename it in checkMintable

File: dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java [hunk 16]
```diff
@@ -67,9 +68,13 @@ public void setProviders(List<IdentifierProvider> providers) {
...
         for (IdentifierProvider service : providers) {
-            String identifier = service.mint(context, dso);
-            if (!StringUtils.isEmpty(identifier)) {
-                service.reserve(context, dso, identifier);
+            try {
+                String identifier = service.mint(context, dso);
+                if (!StringUtils.isEmpty(identifier)) {
+                    service.reserve(context, dso, identifier);
+                }
+            } catch (DOIIdentifierNotApplicableException e) {
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java:75]: as per below comment as well, and the fact that service.reserve() returns void, I think we agree an exception is OK here - shall I refactor it to IdentifierNotApplicable?

File: dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java [hunk 17]
```diff
@@ -82,7 +87,11 @@ public void reserve(Context context, DSpaceObject dso, String identifier)
         // Next resolve all other services
         for (IdentifierProvider service : providers) {
             if (service.supports(identifier)) {
-                service.reserve(context, dso, identifier);
+                try {
+                    service.reserve(context, dso, identifier);
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java:90]: Shall I refactor it to IdentifierNotApplicable?

File: dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java (hunk 12 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java]: thanks @corrad82-4s - sorry for the late reply to this.
i have refactored the method signature and usage here - it is now public void checkMintable() and does not return a value
if it looks good to you, please resolve this conversation

File: dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java (hunk 11 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java]: Hi @LucaGiamminonni , sorry for the late reply to this.
i have refactored LogicalStatementException to be unchecked (by extending RuntimeException instead of Exception)
if it looks good to you, please resolve this conversation

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java (hunk 10 above)
Comment (kshepherd) [reply dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:299]: That is what itemTwo is doing -- the check is for "title begins with TEST", and itemTwo has a title which is expected to not match.

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java (hunk 6 above)
Comment (abollini) [reply dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:81]: ok but better to make this explicit in the javadoc

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java (hunk 10 above)
Comment (abollini) [reply dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:357]: I'm asking to cover an additional use case when the logic filter involve a metadata that is not present in the item at all. This to proof that no NPE are raised in such condition or a false positive match

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java (hunk 9 above)
Comment (abollini) [reply dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:223]: the comment here and in the other test about the fact that the item can be null should be removed

**Review by abollini — CHANGES_REQUESTED** [review]
hi @kshepherd thanks to move this forward.
I have flagged as solved most of my previous comments, the following ones seem to be not addressed or answered so far. Can you double check?

https://github.com/DSpace/DSpace/pull/2789#discussion_r640071356

https://github.com/DSpace/DSpace/pull/2789#discussion_r517870308

https://github.com/DSpace/DSpace/pull/2789#discussion_r640073312

File: dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java (hunk 11 above)
Comment (LucaGiamminonni) [reply dspace-api/src/main/java/org/dspace/content/logic/LogicalStatementException.java]: Hi @kshepherd , sorry for my late reply. So that's ok for me, thank you

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java (hunk 10 above)
Comment (kshepherd) [reply dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:357]: Understood, I've added a further test for an item where the field is "dc.subject" which the item doesn't contain at all

File: dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java (hunk 9 above)
Comment (kshepherd) [reply dspace-api/src/test/java/org/dspace/content/logic/LogicalFilterTest.java:223]: Understood - I fixed the comment and also used itemOne as a test case for all operator tests just to make it clearer that there is an item passed, and avoid any random null handling as you requested

File: dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java (hunk 6 above)
Comment (kshepherd) [reply dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:81]: I had a further think about this and realised primitive is fine - this is the skipFilter boolean anyway, not result (which was already refactored to primitive!) So i've refactored to primitive boolean and tested, we can resolve this.

**Comment by kshepherd** [comment]
@abollini all issues resolved! thanks

**Review by abollini — APPROVED** [review]
it looks fine to me. Thanks for this excellent contribution @kshepherd

**Comment by abollini** [comment]
@tdonohue at the last dev mtg 2021-08-26 we agree to merge once I have reviewed as this code is provided by @kshepherd @pnbecker and was already merged and used in DSpace-CRIS. I will wait until tomorrow to merge just in case you want to give a final look

**Comment by tdonohue** [comment]
@abollini : Feel free to merge... I trust you, @kshepherd and @pnbecker .  I won't have a chance to review this in any great detail, so I'm fine with it moving forward.
